### DISCUSSION TITLE
Command vs Query

### DISCUSSION THREAD

**Discussion by rkqiku** [Dotnet-Boxed/Templates#486] [Q&A]
### Question
Does it make more sense to separate Queries and Commands?
Currently, everything is named as command. 
I mean, is it better to name GET as Queries (GetCarQuery) and POST, PUT etc as Commands (PostCarCommand)?

### Project template
API

## Environment
*

**Comment by RehanSaeed** [comment]
They're named `Command` after the [command pattern](https://en.m.wikipedia.org/wiki/Command_pattern). Commands and queries correspond more with the CQRS pattern which the templates don't implement but they could I suppose.

@VictorioBerra what do you think? I'm not sure if this should be the new default, it does make sense somewhat.

**Comment by VictorioBerra** [comment]
I almost replied to this yesterday but wanted to see your thoughts @RehanSaeed if we are talking about a simple move+rename, I think thats just fine. @rkqiku send a PR.

I checked other CQRS sample repos and notice this pattern: https://github.com/kgrzybek/sample-dotnet-core-cqrs-api/tree/master/src/SampleProject.API/Orders/GetCustomerOrderDetails (`GetCustomerOrderDetailsQuery.cs`)

Alternatively, I see people [make a new folder for Queries](https://github.com/gautema/CQRSlite/tree/master/Sample/CQRSCode/ReadModel/Queries) and keep them separate like that, which seems nice.

**Comment by RehanSaeed** [comment]
I was thinking of a straight rename. If I was going to do subfolders, Wouldn't you prefer to make subfolders based on the controller name:

```
Commands
    Foo
        WooCommand.cs
        WooQuery.cs
    Bar
        WaaCommand.cs
        WaaQuery.cs
Controllers
    FooController.cs
    BarController.cs
```

As opposed to this `Commands` and `Queries` below?

```
Commands
    WooCommand.cs
    WaaCommand.cs
Queries
    WooQuery.cs
    WaaQuery.cs
```

**Comment by VictorioBerra** [comment]
@RehanSaeed there's definitely a trend doing it the first way I have noticed, naming folders by feature. I like the first way but either works!

**Comment by rkqiku** [comment]
Thanks a lot for your feedbacks! I will try my first PR here.
Maybe second one fits better @RehanSaeed 
Every time I discuss separation by Features vs Components this comes into my mind https://pbs.twimg.com/media/DZXxze3XcAA1dbv?format=jpg&name=medium

**Comment by RehanSaeed** [comment]
I thought about this some more. The beginning of the file has the HTTP method in it `Get`, `Post`, `Patch` etc. That provides more information than just `Command` and `Query` and there is some duplication of meaning i.e. a `GET` is a `Query` and everything else is a `Command`. So I'm not too sure it is that useful anymore.

**Comment by disklosr** [comment]
If you're going to do it the CQRS I think there's value in distinguishing commands vs queries at the class name and file name level. Also, I don't think the mapping between http verbs and query/command is a duplication. It sure maps naturally in most cases but it's easy to imagine exceptions to the rule (a POST request mapped to a query)

**Comment by RehanSaeed** [comment]
Sorry for coming back to this after some time. Things got on top of me.

If you go with class name e.g. `GetFooQuery` or `PostFooCommand`. That seems ok. What folder structure would you go with though?

There is a danger that people are going to get confused between the command pattern and CQRS pattern because both use the word 'Command'. Ideally, I'd like to avoid that somehow.
